File: app/src/main/java/com/morsego/app/keyer/IambicKeyerEngine.java
package com.morsego.app.keyer;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

public class IambicKeyerEngine {

    public interface KeyerListener {
        void onToneStart();
        void onToneStop();
        void onElementEmitted(char element);
    }

    private final KeyerSettings settings;
    private final KeyerListener listener;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    private final AtomicBoolean isRunning = new AtomicBoolean(false);
    private volatile boolean ditPressed = false;
    private volatile boolean dahPressed = false;
    private volatile boolean ditMemory = false;
    private volatile boolean dahMemory = false;
    private volatile boolean ditNeedsRelease = false;
    private volatile boolean dahNeedsRelease = false;

    public IambicKeyerEngine(KeyerSettings settings, KeyerListener listener) {
        this.settings = settings;
        this.listener = listener;
    }

    private final Object loopLock = new Object();

    public void onDitChanged(boolean pressed) {
        synchronized (loopLock) {
            ditPressed = pressed;
            if (settings.getMode() == KeyerSettings.Mode.STRAIGHT_KEY) {
                handleStraightKey();
                return;
            }

            if (pressed) {
                ditNeedsRelease = false;
                ditMemory = true;
                startLoopIfNeededLocked();
            } else {
                ditNeedsRelease = false;
            }
        }
    }

    public void onDahChanged(boolean pressed) {
        synchronized (loopLock) {
            dahPressed = pressed;
            if (settings.getMode() == KeyerSettings.Mode.STRAIGHT_KEY) {
                handleStraightKey();
                return;
            }

            if (pressed) {
                dahNeedsRelease = false;
                dahMemory = true;
                startLoopIfNeededLocked();
            } else {
                dahNeedsRelease = false;
            }
        }
    }

    private void handleStraightKey() {
        if (ditPressed || dahPressed) {
            if (listener != null) listener.onToneStart();
        } else {
            if (listener != null) listener.onToneStop();
        }
    }

    private void startLoopIfNeededLocked() {
        if (isRunning.get()) return;
        isRunning.set(true);
        executor.execute(this::runLoop);
    }

    private void runLoop() {
        char lastElement = ' ';

        while (true) {
            boolean curDit;
            boolean curDah;

            synchronized (loopLock) {
                curDit = (!ditNeedsRelease && ditPressed) || ditMemory;
                curDah = (!dahNeedsRelease && dahPressed) || dahMemory;

                if (!curDit && !curDah) {
                    isRunning.set(false);
                    break;
                }
            }

            int wpm = settings.getWpm();
            long ditDuration = MorseTiming.ditDurationMs(wpm);
            long dahDuration = MorseTiming.dahDurationMs(wpm);
            long elementSpace = MorseTiming.intraCharSpaceMs(wpm);

            boolean sendDit;
            if (curDit && curDah) {
                sendDit = (lastElement != '.');
            } else if (curDit) {
                sendDit = true;
            } else {
                sendDit = false;
            }

            if (sendDit) {
                synchronized (loopLock) {
                    ditMemory = false;
                    ditNeedsRelease = true; // Consumed: requires releasing paddle before another dit
                }
                lastElement = '.';
                if (listener != null) {
                    listener.onToneStart();
                    listener.onElementEmitted('.');
                }
                long startDit = System.currentTimeMillis();
                while (System.currentTimeMillis() - startDit < ditDuration) {
                    if (dahPressed) {
                        synchronized (loopLock) {
                            dahMemory = true;
                        }
                    }
                    sleep(2);
                }
                if (listener != null) listener.onToneStop();
            } else {
                synchronized (loopLock) {
                    dahMemory = false;
                    dahNeedsRelease = true; // Consumed: requires releasing paddle before another dah
                }
                lastElement = '-';
                if (listener != null) {
                    listener.onToneStart();
                    listener.onElementEmitted('-');
                }
                long startDah = System.currentTimeMillis();
                while (System.currentTimeMillis() - startDah < dahDuration) {
                    if (ditPressed) {
                        synchronized (loopLock) {
                            ditMemory = true;
                        }
                    }
                    sleep(2);
                }
                if (listener != null) listener.onToneStop();
            }

            // Intra-element spacing (1 dit unit)
            long startPause = System.currentTimeMillis();
            while (System.currentTimeMillis() - startPause < elementSpace) {
                synchronized (loopLock) {
                    if (ditPressed && !ditNeedsRelease) ditMemory = true;
                    if (dahPressed && !dahNeedsRelease) dahMemory = true;
                }
                sleep(2);
            }
        }
    }

    private void sleep(long ms) {
        try {
            Thread.sleep(ms);
        } catch (InterruptedException ignored) {}
    }

    public void stop() {
        synchronized (loopLock) {
            isRunning.set(false);
            ditMemory = false;
            dahMemory = false;
            ditNeedsRelease = false;
            dahNeedsRelease = false;
        }
        if (listener != null) listener.onToneStop();
    }

    public void release() {
        stop();
        executor.shutdownNow();
    }
}

File: app/src/main/java/com/morsego/app/keyer/IambicKeyerEngine.kt
package com.morsego.app.keyer

import com.morsego.app.model.KeyerMode
import com.morsego.app.model.KeyerSettings
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Handles Iambic dual-paddle automatic timing (Mode A, Mode B) and Straight Key manual mode.
 */
class IambicKeyerEngine(
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default),
    private var settings: KeyerSettings = KeyerSettings(),
    private val onToneStart: () -> Unit,
    private val onToneStop: () -> Unit,
    private val onElementSent: (Char) -> Unit
) {
    private var keyerJob: Job? = null

    @Volatile
    private var ditPressed = false

    @Volatile
    private var dahPressed = false

    @Volatile
    private var ditMemory = false

    @Volatile
    private var dahMemory = false

    fun updateSettings(newSettings: KeyerSettings) {
        settings = newSettings
    }

    fun onDitChanged(pressed: Boolean) {
        ditPressed = pressed
        if (pressed && settings.mode != KeyerMode.STRAIGHT_KEY) {
            ditMemory = true
            ensureEngineRunning()
        } else if (settings.mode == KeyerMode.STRAIGHT_KEY) {
            handleStraightKey()
        }
    }

    fun onDahChanged(pressed: Boolean) {
        dahPressed = pressed
        if (pressed && settings.mode != KeyerMode.STRAIGHT_KEY) {
            dahMemory = true
            ensureEngineRunning()
        } else if (settings.mode == KeyerMode.STRAIGHT_KEY) {
            handleStraightKey()
        }
    }

    private fun handleStraightKey() {
        if (ditPressed || dahPressed) {
            onToneStart()
        } else {
            onToneStop()
        }
    }

    private fun ensureEngineRunning() {
        if (keyerJob?.isActive == true) return

        keyerJob = scope.launch {
            var lastElement = ' '

            while (isActive) {
                val currentDit = ditPressed || ditMemory
                val currentDah = dahPressed || dahMemory

                if (!currentDit && !currentDah) {
                    break
                }

                val ditDuration = MorseTiming.ditDurationMs(settings.wpm)
                val dahDuration = MorseTiming.dahDurationMs(settings.wpm)
                val elementSpace = MorseTiming.intraCharSpaceMs(settings.wpm)

                // Decide which element to send next
                val sendDit: Boolean = when {
                    currentDit && currentDah -> {
                        // Squeeze keying alternates
                        lastElement != '.'
                    }
                    currentDit -> true
                    currentDah -> false
                    else -> true
                }

                if (sendDit) {
                    ditMemory = false
                    lastElement = '.'
                    onToneStart()
                    onElementSent('.')
                    delay(ditDuration)
                    onToneStop()
                } else {
                    dahMemory = false
                    lastElement = '-'
                    onToneStart()
                    onElementSent('-')
                    delay(dahDuration)
                    onToneStop()
                }

                // Intra-element spacing (1 Dit)
                // In Iambic B, paddle touches during this pause latch the next element
                val pauseStart = System.currentTimeMillis()
                while (System.currentTimeMillis() - pauseStart < elementSpace) {
                    if (ditPressed) ditMemory = true
                    if (dahPressed) dahMemory = true
                    delay(5)
                }

                // In Iambic A, memory is cleared if paddle was released
                if (settings.mode == KeyerMode.IAMBIC_A) {
                    if (!ditPressed) ditMemory = false
                    if (!dahPressed) dahMemory = false
                }
            }
        }
    }

    fun stop() {
        keyerJob?.cancel()
        ditMemory = false
        dahMemory = false
        onToneStop()
    }
}
